### PR TITLE
Feature/routing  DHT RPC service for single node

### PR DESCRIPTION
**Pull request by auxten**
1. consistent hash data persistence is ok
2. DHT RPC service for single node is ok

### CODE REVIEW

File: crypto/kms/pubkeystore.go [hunk 1]
```diff
@@ -45,23 +45,26 @@ const (
 
 var (
 	// pks holds the singleton instance
-	pks     *PublicKeyStore
-	pksOnce sync.Once
+	pks *PublicKeyStore
+	// for easy test we make PksOnce exported
```
Comment (covenantdb) [inline crypto/kms/pubkeystore.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var PksOnce should be of the form "PksOnce ..."

File: crypto/kms/pubkeystore.go [hunk 2]
```diff
@@ -135,54 +144,163 @@ func GetPublicKey(id proto.NodeID) (publicKey *ec.PublicKey, err error) {
...
+		log.Errorf("get node info failed: %s", err)
+	}
 	return
 }
 
-// SetPublicKey verify nonce and set Public Key
-func SetPublicKey(id proto.NodeID, nonce mine.Uint256, publicKey *ec.PublicKey) (err error) {
-	keyHash := mine.HashBlock(publicKey.SerializeCompressed(), nonce)
-	idHash, err := hash.NewHashFromStr(string(id))
+// GetAllNodeID
```
Comment (covenantdb) [inline crypto/kms/pubkeystore.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported function GetAllNodeID should be of the form "GetAllNodeID ..."

File: crypto/kms/pubkeystore.go [hunk 3]
```diff
@@ -135,54 +144,163 @@ func GetPublicKey(id proto.NodeID) (publicKey *ec.PublicKey, err error) {
...
 	})
+	if err == nil {
+		return (*bolt.DB)(pks.db).Update(func(tx *bolt.Tx) error {
+			bucket := tx.Bucket(pks.bucket)
+			if bucket == nil {
+				return ErrBucketNotInitialized
+			}
+			return bucket.Put([]byte(nodeInfo.ID), nodeBuf)
+		})
+	} else {
```
Comment (covenantdb) [inline crypto/kms/pubkeystore.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
if block ends with a return statement, so drop this else and outdent its block

File: proto/nodeinfo.go [hunk 4]
```diff
@@ -17,34 +17,139 @@
...
+	})
+	if err != nil {
+		log.Errorf("marshal node failed: %s", err)
+		return nil, err
+	}
+
+	return NodeBytes(nodeBuf), err
+}
+
+// Unmarshal the proto.Node with protobuf
```
Comment (covenantdb) [inline proto/nodeinfo.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported function UnmarshalNode should be of the form "UnmarshalNode ..."

File: crypto/kms/pubkeystore.go [hunk 5]
```diff
@@ -135,54 +144,168 @@ func GetPublicKey(id proto.NodeID) (publicKey *ec.PublicKey, err error) {
...
+		log.Errorf("get node info failed: %s", err)
+	}
 	return
 }
 
-// SetPublicKey verify nonce and set Public Key
-func SetPublicKey(id proto.NodeID, nonce mine.Uint256, publicKey *ec.PublicKey) (err error) {
-	keyHash := mine.HashBlock(publicKey.SerializeCompressed(), nonce)
-	idHash, err := hash.NewHashFromStr(string(id))
+// GetAllNodeID
```
Comment (covenantdb) [inline crypto/kms/pubkeystore.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported function GetAllNodeID should be of the form "GetAllNodeID ..."

**Review by zeqing-guo — APPROVED** [review]
LGTM

**Pull request merged**